### PR TITLE
Various submodule fixes

### LINKED ISSUE #823: Calling a function with a sig from within a submodule results in a compiler error
Calling a function with a sig from within a submodule results in a compiler error.
Example:
```clojure
(defmodule Sub
 (defmodule Mod
  (defmodule Ule
    (sig a-func (Fn [String] String))
    (defn a-func [a] a)

    (defn hey [a] (String.append a &(a-func @a))))))

(defn main []
  (println* (Sub.Mod.Ule.hey "Wow")))
```
Result in error:
```
I encountered an error when emitting code:

I found an ambiguous symbol `a-func` at line 7, column 38 in '/Users/tim/example/main.carp'

Possibilities:
  Sub.Mod.Ule.a-func
  Sub.a-func

All possibilities have the correct type.

Traceback:
(build) at Compiler (Build & Run):1:5.
```

Example of working code for context:
```clojure
(defmodule Sub
    (sig a-func (Fn [String] String))
    (defn a-func [a] a)

    (defn hey [a] (String.append a &(a-func @a))))

(defn main []
  (println* (Sub.hey "Wow")))
```
```clojure
(defmodule Sub
 (defmodule Mod
  (defmodule Ule
    ;(sig a-func (Fn [String] String))
    (defn a-func [a] a)

    (defn hey [a] (String.append a &(a-func @a))))))

(defn main []
  (println* (Sub.Mod.Ule.hey "Wow")))
```
```clojure
(defmodule Sub
 (defmodule Mod
   (defmodule Ule
     (defn a-func [a] a)

     (sig hey (Fn [(Ref String)] String))
     (defn hey [a] (String.append a &(a-func @a))))))

(defn main []
 (println* (Sub.Mod.Ule.hey "Wow")))
```

### PR DESCRIPTION
**Pull request by scolsen**
This PR addresses a number of issues with submodule definition/resolution:

In particular it:

- Ensures calling a macro form that generates a binding within a submodule adds the bindings to the correct module (previously they were erroneously hoisted to whatever module was at the toplevel)
- Allows users to define submodules with names of global modules/types without causing a conflict.

fixes #823, #875, #865, #792

Notably, it *doesn't* handle the issue with `use`ing submodules reported in https://github.com/carp-lang/Carp/issues/792#issuecomment-729587870

This also gets us part of the way to #930 -- but we still need to prevent overwriting types in the type environment to get all the way there.

### CODE REVIEW

**Comment by scolsen** [comment]
whew, this one was a huge pain to debug and finally understand what exactly was going on, but it feels good to have figured it out.

I think in general we need to be more explicit w/ lookups in certain contexts. The main problem here was that `lookupInEnv` *always* checks the parent environment, which is typically what we want, but in this case it wasn't since that's what led to the compiler confusing a submodule like `Foo.Dynamic` for the already defined global module `Dynamic`.

Likewise we should make an effort to break down the massive functions w/ tons of local definitions/cases into smaller independent functions. In Eval it was really hard to keep track of what context is being passed where since many of the branches in the main loop just use the original context passed at the top level of the function. If possible, we should try to separate the recursive part of the code from the actual transformations.

**Comment by eriksvedang** [comment]
Wow, closing 4 bugs with one PR! Great job figuring all of this out. 

> If possible, we should try to separate the recursive part of the code from the actual transformations.

Yes, very much agree. Some kind of Context-monad and `traverse` like suggested elsewhere should help a lot, right? 

I've been thinking that the lookup functions should contain a set of already looked up paths to avoid recursion, what do you think about that?

**Comment by scolsen** [comment]
> I've been thinking that the lookup functions should contain a set of already looked up paths to avoid recursion, what do you think about that?

That sounds good!

> Yes, very much agree. Some kind of Context-monad and `traverse` like suggested elsewhere should help a lot, right?

Yes, I think so. Tuples actually already have a monad instance when their left component is a Semigroup. I experimented with defining a semigroup over `Conrtext` so that we can use the monad instance of `(Context, Either EvalError XObj)` but this was kind of awkward. I think our best bet would be to define:

```
deftype EvaluationResult a = Result Context (Either EvalError a)

instance Functor (EvaluationResult a) where
  fmap f (Result ctx res) = Result ctx (fmap f res) 
  
instance Monad (EvalauationResult a) where
  r@(Result ctx res) >>= f = case res of Left _ -> r Right x -> f x
```

etc., where most instances just call the same instance on the interior Either. We can the define custom combinators for running these w/ functions that also want to manipulate ctx.

File: test/vector3.carp [hunk 1]
```diff
@@ -14,7 +14,7 @@
   (assert-op test
              &(init 1.0 2.0 3.0) &(init 1.0 1.0 3.0)
              "/= operator works"
-             Vector3./=)
```
Comment (eriksvedang) [inline test/vector3.carp:17]: Were these changes necessary now with the new env resolution, or is it just left over from testing something?

**Review by eriksvedang — APPROVED** [review]

**Comment by eriksvedang** [comment]
Thanks for providing of the new type you have in mind, I've been pondering how that one should look too but I wasn't sure. If we were to move to a single mutable ref for storing the two root envs, would that change anything. Seems like more stuff would have to be in IO, no?

File: test/vector3.carp (hunk 1 above)
Comment (scolsen) [reply test/vector3.carp:17]: Yes, it is necessary now because of the small lookup resolution changes. I think it’s actually what we want too. Before, because of the way we handled the module lookups, this line found the globally defined `/=` ( from Generics.carp) even though the author here was really trying to reference a function defined in a particular module—I think we want errors in this case otherwise there will be situations in which the resulting behavior is unexpected.

**Comment by scolsen** [comment]
> Thanks for providing an example of the new type you have in mind, I've been pondering how that one should look too, but I wasn't sure. If we were to move to a single mutable ref for storing the two root envs, would that change anything? Seems like more stuff would have to be in IO, no?

It depends exactly what type we use to contain the mutability--it's definitely going to be a monad, but we have some options as to whether we want to "simulate" mutability using a state monad or use IoRef for actual mutability
